pkg/reloader/reloader.go

```diff
@@ -0,0 +1,213 @@
+// Copyright © 2023 Cisco
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package reloader
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"strconv"
+
+	"github.com/bank-vaults/vault-sdk/vault"
+	vaultapi "github.com/hashicorp/vault/api"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	utilruntime "k8s.io/apimachinery/pkg/util/runtime"
+)
+
+func (c *Controller) runReloader(ctx context.Context) {
+	reloaderLogger := c.logger.WithContext(ctx).WithField("worker", "reloader")
+	reloaderLogger.Info("Reloader started")
+
+	if len(c.workloadSecrets.GetWorkloadSecretsMap()) == 0 {
+		reloaderLogger.Info("No secrets to reload")
+		return
+	}
+	if c.vaultClient == nil {
+		reloaderLogger.Info("Vault client is not initialized")
+		c.setVaultClient()
+	}
+
+	// Create a secretWorkloads map and compare the currently used secrets' version
+	// with the one stored in the secretVersions map
+	reloadedWorkloads := make(map[Workload]bool)
+	newSecretVersions := make(map[string]int)
+	for secretPath, workloads := range c.workloadSecrets.GetSecretWorkloadsMap() {
+		reloaderLogger.Debug("Checking secret: ", secretPath)
+		// Get current secret version
+		currentVersion, err := getSecretVersionFromVault(c.vaultClient, secretPath)
+		if err != nil {
+			if err, ok := err.(ErrSecretNotFound); ok {
+				reloaderLogger.Error(err.Error())
+				continue
+			}
+			c.vaultClient = nil
```

Comment:
Any reason why this is being explicitly nilled here?

Reply:
Two things here:
1. Every time `runReloader` is called, we can refresh `vaultClient`. If it fails, we know that issue is related with Vault client itself.
2. When fetching a secret path from vault fails, we can log the error, but skip setting `vaultClient` to nil. This assumes that the error is related to the secret path rather than the client.

Reply:
I also noticed that all `getSecretVersionFromVault` calls will be done serially, which can take quite some time to finish (100 workloads x 10 unique secrets = 1k API calls). For now, we can leave this as is, but it's something to have in mind.

Reply:
10 unique secrets would result in 10 individual calls as we iterate on a secretPath -> workloads map at this point, so only 1 call per secretPath...

Reply:
not if the secrets are unique across all workloads

Reply:
Well yes, if they use a 1000 different secrets, we end up with this many requests, but I don't see any other way as the Vault api doesn't seem to support batch requests: https://github.com/hashicorp/vault/issues/272

Reply:
This was just a side note to have in mind, it's not relevant for the current PR.
We probably won't have to change this, but there is a possible issue that could occur when the reloader loop does not finish before the next trigger. 

P.S. It's possible to mock batch updates via concurrent calls.

Reply:
I see, thanks for the suggestion!

Reply:
Created an issue for it #7